src/Reflex/Backend/Socket/Accept.hs

```diff
@@ -1,123 +1,161 @@
-    getAddr = do
-      addrInfos <- getAddrInfo (Just (defaultHints {addrFlags = [AI_PASSIVE]})) mHost mPort `catch` exHandlerGetAddr
-      pure $ listToMaybe addrInfos
-
-    exHandlerSocket :: IOException -> IO (Maybe Socket)
-    exHandlerSocket e = Nothing <$ onError e
-
-    listenAddr :: AddrInfo -> IO (Maybe Socket)
-    listenAddr h = do
-      sock <- socket (addrFamily h) Stream defaultProtocol
-      bind sock (addrAddress h)
-      listen sock listenQueue
-      pure $ Just sock
-
-    exHandlerAccept :: IOException -> IO (Maybe (Socket, SockAddr))
-    exHandlerAccept e = do
-      void . atomically . tryTakeTMVar $ isOpen
-      Nothing <$ onError e
-
-    acceptLoop :: IO ()
-    acceptLoop = do
-      mSock <- atomically . tryReadTMVar $ isOpen
-      forM_ mSock $ \sock ->  do
-        mConn <- (Just <$> NS.accept sock) `catch` exHandlerAccept
-        forM_ mConn $ \conn -> do
-          onAcceptSocket conn
-          acceptLoop
-
     start = liftIO $ do
-      mAddrInfos <- getAddr
-      forM_ mAddrInfos $ \h -> do
-        mSocket <- listenAddr h `catch` exHandlerSocket
-        forM_ mSocket $ \sock -> do
-          void . atomically . tryPutTMVar isOpen $ sock
-          void . forkIO $ acceptLoop
-
-  performEvent_ $ ffor eClose $ \_ -> liftIO $ do
-    mSock <- atomically . tryTakeTMVar $ isOpen
-    forM_ mSock close
-
-  performEvent_ $ start <$ ePostBuild
-
-  pure $ Accept eAcceptSocket eClosed eError
+      makeListenSocket >>= \case
+        Left exc -> onError exc
+        Right sock -> void . forkIO $ acceptLoop sock
+
+      where
+        makeListenSocket =
+          let
+            getAddrs :: ExceptT IOException IO (NonEmpty AddrInfo)
+            getAddrs = ExceptT . try $
+              -- fromList is probably OK here, as getaddrinfo(3) is required
+              -- to return a nonempty list of addrinfos.
+              --
+              -- See: http://pubs.opengroup.org/onlinepubs/9699919799/functions/getaddrinfo.html
+              -- And: https://github.com/haskell/network/issues/407
+              fromList <$> NS.getAddrInfo (Just passiveHints) mHost mService
+              where passiveHints = NS.defaultHints { addrFlags = [AI_PASSIVE] }
+
+            tryListen :: AddrInfo -> ExceptT IOException IO Socket
+            tryListen info = ExceptT . try $ do
+              sock <- NS.socket (addrFamily info) NS.Stream NS.defaultProtocol
+              (`onException` NS.close sock) $ do
+                NS.setSocketOption sock NS.ReuseAddr 1
+                NS.bind sock (addrAddress info)
+                NS.listen sock listenQueue
+              pure sock
+
+          in runExceptT $ do
+            addrs <- getAddrs
+            let attempts = withExceptT Last . tryListen <$> addrs
+            withExceptT getLast $ asum1 attempts
+
+        acceptLoop sock =
+          let
+            exHandlerAccept :: IOException -> IO (Maybe (Socket, NS.SockAddr))
+            exHandlerAccept e = do
```

Comment:
It is not clear to me that this is the correct behaviour: if the server socket errors once, we should stop accepting connections forever?

Reply:
I am starting to think this is not what we should do: if we fail to accept because we're using too many open files (say), perhaps the correct behaviour is to keep going? Some of the other open connections will go away, and then we can resume normal service.

Reply:
I tried a variant of the server code where errors were signaled without terminating the accept loop, and it ran out of files and hard-locked my system. I think it's okay for us to terminate the accept socket on the first error. A library client who wants a robust server can then build in their own delays, and try to bring the listen socket back up, or whatever.